app/core/agents/interpreter/sandbox_runner.py

```
@@ -0,0 +1,287 @@
+    configure_timeout(int(config["timeout_seconds"]))
+    disable_network()
+    safe_open = patch_filesystem_access(session_dir, config["allowed_input_paths"])
+
+    stdout_buffer = io.StringIO()
+    stderr_buffer = io.StringIO()
+    safe_builtins = build_safe_builtins(safe_open)
+    sandbox_globals = {"__builtins__": safe_builtins, "__name__": "__sandbox__"}
+
+    try:
+        with contextlib.redirect_stdout(stdout_buffer), contextlib.redirect_stderr(stderr_buffer):
+            exec(compile(config["code"], "<metabot-interpreter>", "exec"), sandbox_globals, sandbox_globals)
```

Ruff (0.15.11)

[error] Use of `exec` detected

(S102)